{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Librairie & paramètres\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import plotly.express as px\n",
    "from scipy.cluster import hierarchy\n",
    "import spacy\n",
    "\n",
    "chemin_actuel = os.getcwd()\n",
    "dossier_parent = os.path.abspath(os.path.join(chemin_actuel, os.pardir))\n",
    "dossier_parent = os.path.abspath(os.path.join(dossier_parent, os.pardir))\n",
    "\n",
    "param = {\n",
    "\n",
    "    # Direction vers les fichiers GeoSRA\n",
    "    \"GeoSRA DIR IN\" :dossier_parent+\"/data/raw/geo-sra-ecological.parquet.gzip\",\n",
    "    \"GeoSRA DIR OUT\":dossier_parent+\"/data/processed/geosra_cluster.parquet.gzip\",\n",
    "\n",
    "    # Taille des figures\n",
    "    \"Width Figure\"  :800,\n",
    "    \"Height Figure\" :600,\n",
    "\n",
    "    # Seuils pour les fonctions de répartition et le dendogramme\n",
    "    \"Repart Thr\"    :0.95,\n",
    "    \"Semant Thr\"    :3.5,\n",
    " \n",
    "    # Classe d'assay type principales\n",
    "    \"Assay Class\"   :['AMPLICON','WGS','RNA-Seq','WGA'],\n",
    "\n",
    "    # Couleurs\n",
    "    \"Clusters Color\":{\"(?)\":'lightgrey',\n",
    "                    \"C1\":'darkorange',\n",
    "                    \"C2\":'limegreen', \n",
    "                    \"C3\":'darkred',\n",
    "                    \"C4\":'rebeccapurple',\n",
    "                    \"C5\":'saddlebrown',\n",
    "                    \"C6\":'pink',\n",
    "                    \"C7\":\"gray\",\n",
    "                    \"C8\":\"lightgreen\",\n",
    "                    \"C9\":\"cyan\",\n",
    "                    \"NC\":\"black\"\n",
    "    },\n",
    "\n",
    "    \"Assays Color\"  :{\"(?)\":'lightgrey',\n",
    "                    \"AMPLICON\":\"skyblue\",\n",
    "                    \"WGS\":\"salmon\",\n",
    "                    \"RNA-Seq\":\"lemonchiffon\",\n",
    "                    \"WGA\":\"palegreen\",\n",
    "                    \"OTHER\":\"black\"\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ouvre GeoSRA et affiche les colonnes\n",
    "geosra_pd = pd.read_parquet(param[\"GeoSRA DIR IN\"])\n",
    "geosra_pd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TreeMap des organism\n",
    "\n",
    "organism = list(geosra_pd['organism'].unique())\n",
    "sizes = [len(geosra_pd[geosra_pd['organism'] == o]) for o in organism]\n",
    "\n",
    "treemap_pd = pd.DataFrame(\n",
    "    dict(organism=organism,sizes=sizes)\n",
    ")\n",
    "treemap_pd.loc[treemap_pd[\"organism\"] == '','organism'] = 'Unspecified'\n",
    "\n",
    "treemap_pd[\"all\"] = \"all\"\n",
    "fig = px.treemap(treemap_pd,\n",
    "                path=['all','organism'], \n",
    "                values='sizes',\n",
    ")\n",
    "fig.update_traces(root_color=\"lightgrey\")\n",
    "fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))\n",
    "fig.update_layout(\n",
    "    width=param[\"Width Figure\"],\n",
    "    height=param[\"Height Figure\"]\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fonction de répartition de la colonne organism\n",
    "\n",
    "value_count = dict(geosra_pd[\"organism\"].value_counts())\n",
    "keys   = [i for i in range(len(value_count))]\n",
    "values = list(value_count.values())\n",
    "\n",
    "# Initialize tot and repartiontion list\n",
    "tot    = sum(values)\n",
    "repart = [0]\n",
    "\n",
    "# Repartition for each X\n",
    "for i in range(len(values)-1):\n",
    "    repart.append(sum(values[:i+1])/tot)\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    dict(keys=keys,values=repart)\n",
    ")\n",
    "\n",
    "thr = min([i for i,val in enumerate(repart) if val > param[\"Repart Thr\"]])\n",
    "\n",
    "fig = px.line(df, x=\"keys\", y=\"values\", title='Fonction de répartition')\n",
    "fig.add_vline(x=thr, line_width=3, line_dash=\"dash\", line_color=\"lightcoral\")\n",
    "fig.update_layout(\n",
    "    width=param[\"Width Figure\"],\n",
    "    height=param[\"Height Figure\"]\n",
    ")\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dendogramme (Sémantique) des principales valeurs de organism\n",
    "\n",
    "chains = list(value_count.keys())[:thr]\n",
    "chains = [chain.replace('metagenome', '') for chain in chains]\n",
    "\n",
    "nlp = spacy.load(\"en_core_web_md\")\n",
    "dist_matrix = np.zeros((len(chains), len(chains)))\n",
    "for i in range(len(chains)):\n",
    "    for j in range(len(chains)):\n",
    "        dist_matrix[i, j] = nlp(chains[i]).similarity(nlp(chains[j]))\n",
    "Z = hierarchy.linkage(dist_matrix, method='ward')\n",
    "\n",
    "\n",
    "plt.figure(figsize=(10, 5))\n",
    "dn = hierarchy.dendrogram(Z, labels=chains, color_threshold=param[\"Semant Thr\"])\n",
    "plt.xticks(rotation='vertical')\n",
    "plt.ylabel('Semantix distance')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Histogramme des clusters\n",
    "keys = [key+'metagenome' for key in dn['ivl']]\n",
    "class_dict = dict(zip(keys,dn[\"leaves_color_list\"]))\n",
    "geosra_pd[\"cluster_organism\"] = geosra_pd[\"organism\"].map(class_dict)\n",
    "geosra_pd[\"cluster_organism\"] = geosra_pd[\"cluster_organism\"].fillna('NC')\n",
    "\n",
    "df = geosra_pd[[\"cluster_organism\",\"assay_type\"]]\n",
    "idxs = df[~df['assay_type'].isin(param[\"Assay Class\"])].index\n",
    "df.loc[idxs,\"assay_type\"] = 'OTHER'\n",
    "\n",
    "fig = px.histogram(df, x=\"cluster_organism\", color=\"assay_type\",\n",
    "                color_discrete_map=param[\"Assays Color\"]) \n",
    "fig.update_layout(\n",
    "    width=param[\"Width Figure\"],\n",
    "    height=param[\"Height Figure\"]\n",
    ")            \n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Histogramme des assay types\n",
    "\n",
    "fig = px.histogram(df, x=\"assay_type\", color=\"cluster_organism\", \n",
    "                color_discrete_map=param[\"Clusters Color\"]) \n",
    "fig.update_layout(\n",
    "    width=param[\"Width Figure\"],\n",
    "    height=param[\"Height Figure\"]\n",
    ")            \n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TreeMap des clusters d'organism\n",
    "\n",
    "organism = list(geosra_pd['organism'].unique())\n",
    "cluster = [geosra_pd.loc[geosra_pd['organism'] == o, 'cluster_organism'].values[0] for o in organism]\n",
    "sizes = [len(geosra_pd[geosra_pd['organism'] == o]) for o in organism]\n",
    "\n",
    "treemap_pd = pd.DataFrame(\n",
    "    dict(organism=organism,cluster=cluster,sizes=sizes)\n",
    ")\n",
    "treemap_pd.loc[treemap_pd[\"organism\"] == '','organism'] = 'Unspecified'\n",
    "\n",
    "treemap_pd[\"all\"] = \"all\"\n",
    "fig = px.treemap(treemap_pd,\n",
    "                path=['all','cluster','organism'], \n",
    "                values='sizes',\n",
    "                color='cluster',\n",
    "                color_discrete_map=param[\"Clusters Color\"]\n",
    ")\n",
    "fig.update_traces(root_color=\"lightgrey\")\n",
    "fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))\n",
    "fig.update_layout(\n",
    "    width=param[\"Width Figure\"],\n",
    "    height=param[\"Height Figure\"]\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enregistrement des données avec release_year et cluster_organism\n",
    "geosra_pd.to_parquet(param[\"GeoSRA DIR OUT\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ML",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
